File: 2021/mkday.go
package main

import (
	"fmt"
	"io"
	"log"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"os"
	"path"

	_ "embed"
)

const AOC = "https://adventofcode.com"
const YEAR = "2021"

//go:embed template.txt
var template string

// aocGetRequest puts in the correct cookies to talk with AoC.
func aocGetRequest(path, session string) (*http.Response, error) {

	url := &url.URL{
		Scheme: "https",
		Host:   "adventofcode.com",
		Path:   path,
	}

	req, err := http.NewRequest(http.MethodGet, url.String(), nil)
	if err != nil {
		return nil, err
	}
	req.AddCookie(&http.Cookie{Name: "session", Value: session})

	jar, err := cookiejar.New(nil)
	if err != nil {
		return nil, err
	}
	client := http.Client{
		Jar: jar,
	}
	return client.Do(req)
}

func WriteInput(day, session string) error {

	resp, err := aocGetRequest(path.Join(YEAR, "day", day, "input"), session)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Dump to input.txt
	input, err := os.Create(path.Join(day, "input.txt"))
	if err != nil {
		return err
	}
	_, err = io.Copy(input, resp.Body)
	if err != nil {
		return err
	}
	err = input.Sync()
	if err != nil {
		return err
	}

	return nil
}

func WriteSolutionTemplate(path string) error {
	return os.WriteFile(path, []byte(template), 0644)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("usage: mkday day")
		os.Exit(1)
	}
	day := os.Args[1]

	session := os.Getenv("AOCSESSION")
	if session == "" {
		fmt.Println("Set AOCSESSION in env.")
		os.Exit(1)
	}

	// New directory
	if err := os.Mkdir(day, 0755); err != nil {
		log.Fatal(err)
	}

	// Template .go file
	err := WriteSolutionTemplate(path.Join(day, "solve_test.go"))
	if err != nil {
		log.Fatal(err)
	}

	// Get the input
	err = WriteInput(day, session)
	if err != nil {
		log.Fatal(err)
	}
}

File: 2022/src/bin/mkday.rs
use std::{env, fs, process, sync, time};

use anyhow::Result;
use reqwest::{blocking::Client, cookie::Jar, Url};

const TEMPLATE: &str = include_str!("../../data/template.rs");

fn path_to_input(day: i32) -> std::path::PathBuf {
    std::path::PathBuf::from(format!("./data/input_{:02}.txt", day))
}

fn get_input_file(session: &str, day: i32) -> Result<()> {
    let path = path_to_input(day);
    let mut fh = fs::File::create(path)?;

    let url = format!("https://adventofcode.com/2022/day/{}/input", day).parse::<Url>()?;

    let cookie = format!("session={}", session);
    let jar = Jar::default();
    jar.add_cookie_str(&cookie, &url);

    let client = Client::builder()
        .timeout(time::Duration::from_secs(10))
        .cookie_provider(sync::Arc::new(jar))
        .user_agent("github.com/vikblom/advent-of-code by veblomqvist at gmail dot com")
        .build()?;
    let mut resp = client.get(url).send()?;
    // TODO: Is it necessary to check if status is good?
    if !resp.status().is_success() {
        return Err(anyhow::anyhow!(
            "response indicating error: {}",
            resp.status()
        ));
    }
    resp.copy_to(&mut fh)?;

    Ok(())
}

fn write_template(day: i32) -> Result<()> {
    // let relative = std::path::PathBuf::from("../..")
    //     .join(path_to_input(day))
    //     .into_os_string()
    //     .into_string()
    //     // TODO: Why now Err(anyhow!(...)) here?
    //     .map_err(|_| anyhow!("invalid path"))?;
    let src = TEMPLATE.replace("%%DAY%%", &format!("{:02}", day));
    fs::write(format!("./src/bin/day{:02}.rs", day), src)?;
    Ok(())
}

fn main() -> Result<()> {
    let args: Vec<String> = env::args().skip(1).collect();
    if args.len() != 1 {
        println!("usage: mkday <day>");
        process::exit(1);
    }
    let day = args[0].parse::<i32>()?;
    let session = env::var("AOC_SESSION")?;
    // input
    get_input_file(&session, day)?;
    // src
    write_template(day)?;
    Ok(())
}
